src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -285,16 +285,44 @@ protected virtual void Unlock()
 				{
 					// Cloned transaction is not disposed "unexpectedly", its status is accessible till context disposal.
 					var status = EnlistedTransaction.TransactionInformation.Status;
-					if (status != TransactionStatus.Active)
+					if (status != TransactionStatus.Active || _preparing)
 						return status;
 
-					// The clone status can be out of date when active, check the original one (which could be disposed if
-					// the clone is out of date).
+					// The clone status can be out of date when active and not in prepare phase, in case of rollback or
+					// dependent clone usage.
+					// In such case the original transaction is already disposed, and trying to check its status will
+					// trigger a dispose exception.
 					return _originalTransaction.TransactionInformation.Status;
 				}
 				catch (ObjectDisposedException ode)
 				{
-					_logger.Warn(ode, "Enlisted transaction status was wrongly active, original transaction being already disposed. Will assume neither active nor committed.");
+					// For ruling out the dependent clone case when possible, we check if the current transaction is
+					// equal to the context one (System.Transactions.Transaction does override equality for this), and
+					// in such case, we check the state of the current transaction instead. (The state of the current
+					// transaction if equal can only be the same, but it will be inaccessible in case of rollback, due
+					// to the current transaction being already disposed.)
+					// The current transaction may not be reachable during 2PC phases and transaction completion events,
+					// but in such cases the context transaction is either no more active or in prepare phase, which is
+					// already covered by _preparing test.
+					try
+					{
+						var currentTransaction = System.Transactions.Transaction.Current;
+						if (!ReferenceEquals(currentTransaction, _originalTransaction) &&
+							currentTransaction == EnlistedTransaction)
+							return currentTransaction.TransactionInformation.Status;
```

Comment:
This allows chaining multiple dependent transaction usage of the same original transaction on the same session. This means this use case will works with some exception handling, which hurts performances.
But dodging theses exceptions for this dependent case would require to check the current transaction first (instead of trying the original one), which would introduce another exception handling for more regular cases than dependent transaction or rollback.